src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,820 @@
+// In particular, len(), adjust(), trim(), clone(), and split() are present and designed to behave the same regardless.
+//
+// The constructors/destructors, however, are necessarily different.  For DPDK-allocated buffers, a MBuf is expected
+// to be allocated externally and provided to the DemiBuffer's "from_mbuf" constructor.  A MBuf can also be extracted
+// from a DPDK allocated DemiBuffer via the "into_mbuf" routine.
+//
+// Note: if compiled without the "libdpdk" feature defined, the DPDK-specific functionality won't be present.
+
+// Note on buffer chain support:
+// DPDK has a concept of MBuf chaining where multiple MBufs may be linked together to form a "packet".  While the
+// DemiBuffer routines for heap-allocated buffers also now support this functionality, it isn't yet exposed via the
+// DemiBuffer interface.
+// ToDo: Expose buffer chain support once we have a solid use case.
+
+// Note on intrusive queueing:
+// Since all DemiBuffer types keep the metadata for each "view" in a separate allocated region, they can be queued
+// using intrusive links (i.e. have a link field in the metadata).
+// ToDo: Expose calls to get/set a linking field.
+
+// Note on the allocation functions:
+// This code currently uses std::alloc() and std::dealloc() to allocate/free things from the heap.  Note that the Rust
+// documentation says that these functions are expected to be deprecated in favor of their respective methods of the
+// "Global" type when it and the "Allocator" trait become stable.
+use crate::runtime::fail::Fail;
+#[cfg(feature = "libdpdk")]
+use ::dpdk_rs::{
+    rte_mbuf,
+    rte_mempool,
+    rte_pktmbuf_adj,
+    rte_pktmbuf_clone,
+    rte_pktmbuf_free,
+    rte_pktmbuf_trim,
+};
+#[cfg(feature = "libdpdk")]
+use ::std::mem;
+use ::std::{
+    alloc::{
+        alloc,
+        dealloc,
+        handle_alloc_error,
+        Layout,
+    },
+    marker::PhantomData,
+    mem::size_of,
+    num::NonZeroUsize,
+    ops::{
+        Deref,
+        DerefMut,
+    },
+    ptr::{
+        self,
+        null_mut,
+        NonNull,
+    },
+    slice,
+};
+
+// Cache line size, defined here for code clarity.
+// While this is arguably architecture dependent, and thus should be pulled from some architecture-specific information,
+// this value (64 bytes) is used by all architectures we're likely to care about (and is also assumed by DPDK).
+const CACHE_LINE_SIZE: usize = 64;
+
+// Buffer Metadata.
+// This is defined to match a DPDK MBuf (rte_mbuf) in order to potentially use the same code for some DemiBuffer
+// operations that currently use identical (but separate) implementations for heap vs DPDK allocated buffers.
+// Fields beginning with an underscore are not directly used by the current DemiBuffer implementation.
+// Should be cache-line aligned and consume 2 cache lines (128 bytes).
+#[repr(C)]
+#[repr(align(64))]
+struct MetaData {
```

Comment:
Could you add an assertion on this size?

Reply:
See Line 532.

Reply:
Thanks for pointing me to this line. I do have additional points:

- This shouldn't be a debug assert, since it breaks functionality
- This check should be done as soon as possible at initialization time (Demikernel initialization should fail)

Reply:
It actually doesn't break functionality, it just breaks efficiency, so I think a debug assert is fine.

This whole thing is just a work-around for Rust's (hopefully temporary) lack of proper compile-time assertions (which is how I would have done this in C).  Instead, I check the size at first use of the struct, and only in debug builds so we don't take the hit of the assertion check in production on every allocation call.  I do think the check for this should be co-located with the struct definition rather than in some other unrelated part of the code.

I see only three reasonable courses of action here:
1. Leave it as is.
1. Use one of the [hacky](https://www.reddit.com/r/rust/comments/9z8iqa/compile_time_assertions_trick/) [ways](https://crates.io/crates/static_assertions) to fake compile-time assertions.
1. Remove the check entirely.

Reply:
4. Use this:

```rust
macro_rules! assert_size_eq {
    ($x:ty, $($xs:ty),+ $(,)?) => {
        const _: fn() = || {
            $(let _ = $crate::_core::mem::transmute::<$x, $xs>;)+
        };
    };
}
```

Reply:
Good news!  I found out that Rust now provides a less hacky way to have compile-time assertions.  So, I'm going to switch to that.

Reply:
Added two compile-time assertions regarding the MetaData struct:
- Check that MetaData is aligned to a cache line.
- Check that MetaData is two cache lines in size.

And removed the old run-time debug assertion.

Reply:
Awesome! Thanks